### DISCUSSION TITLE
Using MicroPython with 16 MB of flash

### DISCUSSION THREAD

**Discussion by Allan196** [micropython/micropython#9911] [Using MicroPython]
My initial hardware was an ESP32 module with 4MB of flash memory.
I started code development with MicroPython (esp32-20220618-v1.19.1.bin) and Thonny 3.3.13 as the development environment.
All went well but as the code grew larger I started getting these errors:

MemoryError: memory allocation failed, allocating 3070 bytes
>>> 
NB! Thonny could not execute 'get_globals'.

I assumed that I could overcome this but changing to another ESP32 module but with 16 MB of flash.
This is the 16 MB ESP32 module
https://www.sparkfun.com/products/15663

The first thing I did was to use Thonny to run:
>>> esp.flash_size()
The response was 16777216 ie 16 MB as expected.
But the MemoryError remains.

Does MicroPython know how to use the increase in flash memory?

I don't know how to proceed.
I am running Thonny with a Windows 10 PC and don't have access to a Linux PC
Any help appreciated.

**Comment by robert-hh** [comment]
The memory error is about RAM, not flash. So you either need a device with more RAM, like one using a WROVER module, not the WROOM module. Or you could freeze your code into flash, freeing RAM for data. That requires, that you build the firmware yourself.
In the unlikely event that the error happens during import of the code, you could try to pre-compile the code using mpy-cross. Then the RAM usage during import is lower.

**Reply by Allan196** [reply]
Before using MicroPython I have been using PIC microcontrollers where code resides in flash and variables in RAM. Is it the same when using MicroPython for ESP32 development?
I was able to re-write some of the code as modules and copy them to the ESP32 flash (using Thonny) as .py files. Is that different to " freeze your code into flash, freeing RAM for data" as you say?
Thanks for your help.
Allan

**Reply by Allan196** [reply]
I have now upgraded to a Pycom WiPy 3.0. This has 4MB of RAM compared to the ESP32 WROOM with 520 KB of on-chip SRAM.
I installed the same MicroPython (esp32-20220618-v1.19.1.bin) firmware and loaded my same MicroPython code.
Yet I still get MemoryError: memory allocation failed, allocating 1784 bytes

It seems that MicroPython does not recognise the increased RAM as the following result of mem_info() is similar to the ESP32 WROOM with 520 KB of SRAM.
>>> import micropython
>>> micropython.mem_info()
stack: 736 out of 15360
GC: total: 111168, used: 18752, free: 92416
 No. of 1-blocks: 454, 2-blocks: 61, max blk sz: 24, max free sz: 5765

**Reply by robert-hh** [reply]
The Generic firmware does not use the additional RAM. Please try the GENERIC SPIRAM version, e.g. one for ESP32 Wrover devices. https://micropython.org/download/esp32spiram/

**Comment by robert-hh** [comment]
That is different. When you copy a .py source file to the board, it is stored in the board's file system. When you run it, it will be compiled and the compiled to bytecode, placed in RAM and executed from there, using RAM for the data as well. Frozen bytecode will be stored in flash and executed from flash. But embedding frozen bytecode happens, when a firmware is built. That may be similar to your PIC scenario, where you had that build & flash step as well. The Arduino IDE does that.
